[wip] data_import: Add Microsoft Teams importer.

PieterCK opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
Hmm, the changes are a lot more than I initially thought, mainly because we need to call the API to get the user roles. Cutting the fixture files, it's roughly ~1600 LOC. So alternatively, I think this can also be separated into 3 separate PRs. The PR arrangement could probably be something like this:
- First PR: Add `do_convert_directory` and the version that only converts users
- Second PR: Support converting teams → Zulip channels
- Third PR: Support converting teams messages. This PR also adds the documentation and the `convert_microsoft_teams_data` command.

Let me know if we want to switch this to prep PRs.

I've also written down some follow-ups / features that aren't in version 1 in #36295. Let me know if any of those need to be in version 1. The priorities for those follow-ups also need to be confirmed.

Will go through this PR one more time the first thing next week / weekend before marking this as ready for review. Posting this comment earlier to get early inputs on those points.

alya commented:
I think it's fine to stage it this way; we can decide how we want to document the importer at each stage (e.g., we may wait until the High priority features are done before adding it to the help center).

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -355,6 +375,202 @@ def convert_users(
+                display_name=team_channel["DisplayName"],
+                is_favourite_by_default=team_channel["IsFavoriteByDefault"],
+                is_archived=team_channel["IsArchived"],
+                is_favorite_by_deafult=team_channel["IsFavoriteByDefault"],
```
Fixed the typo

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -355,6 +375,202 @@ def convert_users(
+        team_channels_list = get_data_file(
+            os.path.join(team_data_folder, f"channels_{team_id}.json")
+        )
+        for team_channel in team_channels_list:
```
> treated more like Zulip "Channel folders"?

A big merit of this approach is that we can handle the "public team with private channel" situation nicely, all private team's channel will be Zulip private channels and public team's private channel is also Zulip private channel.

I guess the main reason I went with converting Teams team → Zulip channel and Teams channel → Zulip topics is because I'm not sure how well some of the settings would map out.

- The big thing is probably subscribers. Each Teams team potentially wants to keep all new public channels in the channel folder to have the same subscribers (in Teams, all team members are subscribed to public channels in a team), so they would have to manually subscribe team members each time.

- Teams team settings like "Allow members to create and update channels" and "Allow guests to create and update channels" can also be mapped to Zulip channel's "Who can start new topics" setting, but not to a channel folder.

- Teams channels have very few settings, so they’re probably the most compatible to be a Zulip topic. The only lossy conversion is the Teams channel's "who can post messages" setting, which is something Zulip topics don't manage.

All in all, I feel like the current approach is probably the less lossy one and creates less friction for users after import.

PieterCK replied:
More details on MS Teams teams, channels, and messages:

### Team:
- An organization can have multiple teams
- Can be private or public
- Have members and owners
- Contains "channels"
- Team settings (only relevant ones):
  <img width="1281" height="570" alt="image" src="https://github.com/user-attachments/assets/53356418-f88f-4896-a109-5cb1a229b5ea" />
  <img width="1281" height="176" alt="image" src="https://github.com/user-attachments/assets/e48d26a1-eecc-4aae-b06d-ef004e78df59" />

### Channel:
- A team can have multiple channels
- Can be private or public
- Can have members if it's a private channel, otherwise all team members are subscribed to it.
- Contains messages.
- Public channel settings: 
  <img width="728" height="247" alt="image" src="https://github.com/user-attachments/assets/53fb45d1-1b93-4dfb-8ada-7052c7b0ab87" />

- Private channel settings:
  <img width="699" height="416" alt="image" src="https://github.com/user-attachments/assets/af176a77-c034-4b9e-9495-526f2fdc0a8a" />

- Lastly, there is the channel **Layout** setting: 
  <img width="679" height="387" alt="image" src="https://github.com/user-attachments/assets/86a1fc40-3e62-4ee8-8e0c-59234ad09f33" />
    - The **Threads** layout is akin to Slack's threads
    - The **Posts** layout is closer to Facebook posts
    - A channel's layout setting is not exported, it is also not included in the API. 
    - A "post" message and a "thread" message are indistinguishable once exported. 
    - A message that is not a reply to another message will be converted to a post if the "post" layout is chosen, and a normal chat message if it's a "thread" layout.
  
      | Threads | Posts |
      |--------|--------|
      | <img width="1294" height="674" alt="image" src="https://github.com/user-attachments/assets/7b4e6c83-93d5-4996-a955-a869d589e42a" />| <img width="1291" height="690" alt="image" src="https://github.com/user-attachments/assets/20df8343-bae1-405b-bf8c-52f03a8fbbd3" />|

timabbott replied:
I think we can stick with your current approach, and think about changing it in the future (or making it an option) if we get feedback suggesting an alternative approach is better.

alya replied:
Seems reasonable to me.

We could potentially use #26944 for the "only owners can post" setting once we have it.

alya replied:
I think also regardless of how exactly teams/channels were used in Teams, it can be nice to have all imported messages be fairly contained. An org might choose to create new channels, etc., and use the imported content just as a reference.

timabbott commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -341,6 +361,202 @@ def convert_users(
+                raise AssertionError("Failed to build channel data.")
+            current_channel = channel_metadata[message["ChannelIdentity"]["ChannelId"]]
+            if current_channel.membership_type == "private":
+                # Carefule not to include private channel messages.
```
typo.

PieterCK commented on zerver/data_import/microsoft_teams.py:
```diff
@@ -341,6 +361,202 @@ def convert_users(
+    dump_file_id = 1
+    for raw_message_chunk in get_batched_export_message_data(message_file_paths, chunk_size):
+        # Sort Microsoft Teams messages by their ID, which is their date sent in unix time.
+        sorted_message_chunk = sorted(raw_message_chunk, key=lambda m: int(m["Id"]))
```
Not sure if implementing something like a k-way merge to sort all the batched messages into one sorted list is worth all the extra complexity considering `get_batched_export_message_data`'s batching isn't actually being careful about memory usage, it's loading whole message files into memory and batch it after.

PieterCK commented:
I think I've addressed all comments. Highlighting [this](https://github.com/zulip/zulip/pull/36122#discussion_r2658270102) issue for Tim.

@zulipbot remove "maintainer review"
@zulipbot add "integration review"
